File: src/types/agency.ts
export interface ApiEndpoint {
  method: string;
  path: string;
  description: string;
}

export interface Agency {
  id: string;
  name: string;
  shortName: string;
  logo: string;
  connectionType: 'MCP' | 'API' | 'A2A';
  status: 'active' | 'inactive';
  description: string;
  dataScope: string[];
  totalCalls: number;
  color: string;
  endpointUrl: string;
  apiKeyName?: string | null;
  authMethod?: string;
  authHeader?: string;
  basePath?: string;
  rateLimitRpm?: number | null;
  requestFormat?: string;
  apiEndpoints?: ApiEndpoint[];
  apiSpecRaw?: string | null;
  createdAt?: string;
  updatedAt?: string;
}

// DB row shape (snake_case) → mapped to Agency (camelCase)
export interface AgencyRow {
  id: string;
  name: string;
  short_name: string;
  logo: string;
  connection_type: string;
  status: string;
  description: string;
  data_scope: string[];
  total_calls: number;
  color: string;
  endpoint_url: string;
  api_key_name: string | null;
  auth_method: string;
  auth_header: string;
  base_path: string;
  rate_limit_rpm: number | null;
  request_format: string;
  api_endpoints: ApiEndpoint[];
  api_spec_raw: string | null;
  created_at: string;
  updated_at: string;
}

export function mapRowToAgency(row: AgencyRow): Agency {
  return {
    id: row.id,
    name: row.name,
    shortName: row.short_name,
    logo: row.logo,
    connectionType: row.connection_type as Agency['connectionType'],
    status: row.status as Agency['status'],
    description: row.description,
    dataScope: row.data_scope,
    totalCalls: row.total_calls,
    color: row.color,
    endpointUrl: row.endpoint_url,
    apiKeyName: row.api_key_name,
    authMethod: row.auth_method,
    authHeader: row.auth_header,
    basePath: row.base_path,
    rateLimitRpm: row.rate_limit_rpm,
    requestFormat: row.request_format,
    apiEndpoints: row.api_endpoints,
    apiSpecRaw: row.api_spec_raw,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export interface AgentStep {
  icon: string;
  label: string;
  detail?: string;
  status: 'pending' | 'active' | 'done';
}

File: backend/app/schemas/agency.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Sub-schemas
# ---------------------------------------------------------------------------

class ApiEndpoint(BaseModel):
    method: str = Field(..., examples=["GET", "POST"])
    path: str = Field(..., examples=["/search"])
    description: str = Field(default="")


class ResponseField(BaseModel):
    field: str
    type: str = Field(..., examples=["string", "number", "boolean"])
    description: str = Field(default="")
    example: Any = None

class ApiHeader(BaseModel):
    name: str
    value: str
    description: str = Field(default="")


# ---------------------------------------------------------------------------
# Agency schemas
# ---------------------------------------------------------------------------

class AgencyHealthEmbed(BaseModel):
    state: str  # up | degraded | down | unknown
    uptime_24h: float | None = None
    avg_latency_ms_24h: int | None = None
    last_check_at: datetime | None = None


class AgencyBase(BaseModel):
    name: str
    short_name: str | None = None
    logo: str | None = None
    description: str | None = None
    connection_type: str = "API"
    status: str = "active"
    data_scope: list[str] = []
    color: str | None = None

    # Connection
    endpoint_url: str | None = None
    auth_method: str | None = None
    auth_header: str | None = None
    base_path: str | None = None
    api_key_name: str | None = None
    rate_limit_rpm: int | None = None
    request_format: str | None = None

    # Schema / spec
    api_endpoints: list[ApiEndpoint] = []
    response_schema: list[ResponseField] = []
    api_spec_raw: str | None = None
    expected_payload: dict[str, Any] | None = None
    api_headers: list[ApiHeader] | None = None

    # Routing controls
    priority: int | None = None
    router_hint: str = ""
    dispatch_timeout_s: int | None = None
    mcp_tool_name: str | None = None


class AgencyCreate(AgencyBase):
    """Request body for creating a new agency."""
    pass


class AgencyUpdate(BaseModel):
    """Request body for partial update of an agency (all fields optional)."""
    name: str | None = None
    short_name: str | None = None
    logo: str | None = None
    description: str | None = None
    connection_type: str | None = None
    status: str | None = None
    data_scope: list[str] | None = None
    color: str | None = None
    endpoint_url: str | None = None
    auth_method: str | None = None
    auth_header: str | None = None
    base_path: str | None = None
    api_key_name: str | None = None
    rate_limit_rpm: int | None = None
    request_format: str | None = None
    api_endpoints: list[ApiEndpoint] | None = None
    response_schema: list[ResponseField] | None = None
    api_spec_raw: str | None = None
    expected_payload: dict[str, Any] | None = None
    api_headers: list[ApiHeader] | None = None
    priority: int | None = None
    router_hint: str | None = None
    dispatch_timeout_s: int | None = None
    mcp_tool_name: str | None = None


class AgencyResponse(AgencyBase):
    """Response schema — includes server-generated fields."""
    id: uuid.UUID
    total_calls: int
    created_at: datetime
    updated_at: datetime
    rating_up: int = 0
    rating_down: int = 0
    health: AgencyHealthEmbed | None = None

    model_config = ConfigDict(from_attributes=True)


class AgencyListResponse(BaseModel):
    """Paginated list of agencies."""
    data: list[AgencyResponse]
    total: int


class HealthHistoryBucket(BaseModel):
    bucket_start: datetime
    uptime_pct: float
    avg_latency_ms: int
    checks: int
    failures: int


class HealthHistoryResponse(BaseModel):
    data: list[HealthHistoryBucket]


class StatusUpdateRequest(BaseModel):
    status: str


class McpDiscoverRequest(BaseModel):
    endpoint_url: str


class McpToolInfo(BaseModel):
    name: str
    description: str = ""
    input_schema: dict[str, Any] = Field(default_factory=dict)


class McpDiscoverResponse(BaseModel):
    tools: list[McpToolInfo]


# ---------------------------------------------------------------------------
# Summary schema used by MCP resource (lightweight)
# ---------------------------------------------------------------------------

class AgencySummary(BaseModel):
    id: uuid.UUID
    name: str
    short_name: str | None
    logo: str | None
    connection_type: str
    status: str
    data_scope: list[str]
    total_calls: int
    color: str | None

    model_config = ConfigDict(from_attributes=True)
